File: 25. Reverse Nodes in k-Group/Solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        def reverse(a: ListNode, b: ListNode):
            pre = None
            cur = a
            nxt = a
            while cur != b :
                nxt = cur.next
                cur.next = pre
                pre = cur
                cur = nxt
            return pre
        
        if head == None: return None
        a = b = head
        for _ in range(k):
            if b == None:
                return head
            b = b.next
        newHead = reverse(a, b)
        a.next = self.reverseKGroup(b, k)
        return newHead

File: 25. Reverse Nodes in k-Group/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        if (head == null) {return null;}
        ListNode a, b;
        a = b = head;
        for (int i = 0; i < k; i++) {
            if (b == null) {
                return a;
            } 
//         The order of these two lines defines the case [1,2] k = 2
            b = b.next;
             
        }
        ListNode newHead = reverse(a, b);
        a.next = reverseKGroup(b, k);
        return newHead;
    }
    
    public ListNode reverse(ListNode a, ListNode b) {
        ListNode pre, cur, nxt;
        pre = null;
        cur = a;
        nxt = a;
        while (cur != b) {
            nxt = cur.next;
            cur.next = pre;
            pre = cur;
            cur = nxt;
        }
        return pre;
    }
}